File: app/src/main/java/wangdaye/com/geometricweather/settings/adapters/WeatherIconAdapter.java
package wangdaye.com.geometricweather.settings.adapters;

import android.graphics.drawable.Drawable;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.appcompat.widget.AppCompatImageView;
import androidx.recyclerview.widget.GridLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import java.util.List;

import wangdaye.com.geometricweather.R;
import wangdaye.com.geometricweather.basic.GeoActivity;

public class WeatherIconAdapter extends RecyclerView.Adapter<RecyclerView.ViewHolder> {

    private final GeoActivity mActivity;
    private final List<Item> mItemList;

    // item.

    public interface Item {}

    public static class Title implements Item {

        String content;

        public Title(String content) {
            this.content = content;
        }
    }

    public static abstract class WeatherIcon implements Item {

        public abstract Drawable getDrawable();
        public abstract String getContentDescription();

        public abstract void onItemClicked(GeoActivity activity);
    }

    public static class Line implements Item {}

    // holder.

    class TitleHolder extends RecyclerView.ViewHolder {

        private final TextView mTitle;

        TitleHolder(@NonNull View itemView) {
            super(itemView);
            mTitle = itemView.findViewById(R.id.item_weather_icon_title);
        }

        void onBindView() {
            Title t = (Title) mItemList.get(getAdapterPosition());
            mTitle.setText(t.content);
        }
    }

    class IconHolder extends RecyclerView.ViewHolder {

        private final AppCompatImageView mImageView;

        IconHolder(@NonNull View itemView) {
            super(itemView);
            mImageView = itemView.findViewById(R.id.item_weather_icon_image);
        }

        void onBindView() {
            if (getAdapterPosition() == RecyclerView.NO_POSITION) {
                return;
            }
            WeatherIcon icon = (WeatherIcon) mItemList.get(getAdapterPosition());
            mImageView.setImageDrawable(icon.getDrawable());

            itemView.setContentDescription(icon.getContentDescription());
            itemView.setOnClickListener(v -> icon.onItemClicked(mActivity));
        }
    }

    class LineHolder extends RecyclerView.ViewHolder {

        LineHolder(@NonNull View itemView) {
            super(itemView);
        }
    }

    // adapter.

    public WeatherIconAdapter(GeoActivity activity, List<Item> itemList) {
        mActivity = activity;
        mItemList = itemList;
    }

    @NonNull
    @Override
    public RecyclerView.ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        if (viewType == 1) {
            return new TitleHolder(
                    LayoutInflater.from(parent.getContext())
                            .inflate(R.layout.item_weather_icon_title, parent, false)
            );
        }
        if (viewType == -1) {
            return new LineHolder(
                    LayoutInflater.from(parent.getContext())
                            .inflate(R.layout.item_line, parent, false)
            );
        }
        return new IconHolder(
                LayoutInflater.from(parent.getContext())
                        .inflate(R.layout.item_weather_icon, parent, false)
        );
    }

    @Override
    public void onBindViewHolder(@NonNull RecyclerView.ViewHolder holder, int position) {
        if (holder instanceof LineHolder) {
            return;
        }
        if (holder instanceof TitleHolder) {
            ((TitleHolder) holder).onBindView();
        } else {
            ((IconHolder) holder).onBindView();
        }
    }

    @Override
    public int getItemCount() {
        return mItemList.size();
    }

    @Override
    public int getItemViewType(int position) {
        if (mItemList.get(position) instanceof Title) {
            return 1;
        }
        if (mItemList.get(position) instanceof Line) {
            return -1;
        }
        return 0;
    }

    public static GridLayoutManager.SpanSizeLookup getSpanSizeLookup(int columnCount,
                                                                     List<WeatherIconAdapter.Item> itemList) {
        return new SpanSizeLookup(columnCount, itemList);
    }
}

class SpanSizeLookup extends GridLayoutManager.SpanSizeLookup {

    private final int mColumnCount;
    private final List<WeatherIconAdapter.Item> mItemList;

    SpanSizeLookup(int columnCount, List<WeatherIconAdapter.Item> itemList) {
        mColumnCount = columnCount;
        mItemList = itemList;
    }

    @Override
    public int getSpanSize(int position) {
        if (mItemList.get(position) instanceof WeatherIconAdapter.Title
                || mItemList.get(position) instanceof WeatherIconAdapter.Line) {
            return mColumnCount;
        }
        return 1;
    }
}

File: app/src/main/kotlin/org/breezyweather/ui/settings/adapters/WeatherIconAdapter.kt
/*
 * This file is part of Breezy Weather.
 *
 * Breezy Weather is free software: you can redistribute it and/or modify it
 * under the terms of the GNU Lesser General Public License as published by the
 * Free Software Foundation, version 3 of the License.
 *
 * Breezy Weather is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
 * or FITNESS FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public
 * License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with Breezy Weather. If not, see <https://www.gnu.org/licenses/>.
 */

package org.breezyweather.ui.settings.adapters

import android.graphics.drawable.Drawable
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import androidx.appcompat.widget.AppCompatImageView
import androidx.recyclerview.widget.GridLayoutManager
import androidx.recyclerview.widget.RecyclerView
import org.breezyweather.R
import org.breezyweather.common.activities.BreezyActivity

class WeatherIconAdapter(
    private val mActivity: BreezyActivity,
    private val mItemList: List<Item>,
) : RecyclerView.Adapter<RecyclerView.ViewHolder>() {
    // item.
    interface Item
    class Title(var content: String) : Item
    abstract class WeatherIcon : Item {
        abstract val drawable: Drawable
        abstract val contentDescription: String
        abstract fun onItemClicked(activity: BreezyActivity)
    }

    class Line : Item

    // holder.
    internal inner class TitleHolder(
        itemView: View,
    ) : RecyclerView.ViewHolder(itemView) {
        private val mTitle: TextView = itemView.findViewById(R.id.item_weather_icon_title)

        fun onBindView() {
            val t = mItemList[bindingAdapterPosition] as Title
            mTitle.text = t.content
        }
    }

    internal inner class IconHolder(
        itemView: View,
    ) : RecyclerView.ViewHolder(itemView) {
        private val mImageView: AppCompatImageView = itemView.findViewById(R.id.item_weather_icon_image)

        fun onBindView() {
            if (bindingAdapterPosition == RecyclerView.NO_POSITION) return
            val icon = mItemList[bindingAdapterPosition] as WeatherIcon
            mImageView.setImageDrawable(icon.drawable)
            itemView.contentDescription = icon.contentDescription
            itemView.setOnClickListener { icon.onItemClicked(mActivity) }
        }
    }

    internal inner class LineHolder(
        itemView: View,
    ) : RecyclerView.ViewHolder(itemView)

    override fun onCreateViewHolder(
        parent: ViewGroup,
        viewType: Int,
    ): RecyclerView.ViewHolder {
        if (viewType == 1) {
            return TitleHolder(
                LayoutInflater.from(parent.context).inflate(R.layout.item_weather_icon_title, parent, false)
            )
        }
        return if (viewType == -1) {
            LineHolder(
                LayoutInflater.from(parent.context).inflate(R.layout.item_line, parent, false)
            )
        } else {
            IconHolder(
                LayoutInflater.from(parent.context).inflate(R.layout.item_weather_icon, parent, false)
            )
        }
    }

    override fun onBindViewHolder(
        holder: RecyclerView.ViewHolder,
        position: Int,
    ) {
        if (holder is LineHolder) return
        if (holder is TitleHolder) {
            holder.onBindView()
        } else {
            (holder as IconHolder).onBindView()
        }
    }

    override fun getItemCount(): Int {
        return mItemList.size
    }

    override fun getItemViewType(position: Int): Int {
        if (mItemList[position] is Title) return 1
        return if (mItemList[position] is Line) -1 else 0
    }

    companion object {
        fun getSpanSizeLookup(
            columnCount: Int,
            itemList: List<Item>,
        ): GridLayoutManager.SpanSizeLookup {
            return SpanSizeLookup(columnCount, itemList)
        }
    }
}

internal class SpanSizeLookup(
    private val mColumnCount: Int,
    private val mItemList: List<WeatherIconAdapter.Item>,
) : GridLayoutManager.SpanSizeLookup() {
    override fun getSpanSize(position: Int): Int {
        return if (mItemList[position] is WeatherIconAdapter.Title || mItemList[position] is WeatherIconAdapter.Line) {
            mColumnCount
        } else {
            1
        }
    }
}
